canister/src/unstable_blocks.rs

```diff
@@ -71,6 +77,38 @@ impl UnstableBlocks {
     pub fn blocks_with_depths_by_heights(&self) -> Vec<Vec<(&Block, u32)>> {
         self.tree.blocks_with_depths_by_heights()
     }
+
+    /// Returns depth in BlockTree of Block with given BlockHash.
+    fn block_depth(&mut self, block_hash: &BlockHash) -> Result<u32, BlockDoesNotExtendTree> {
+        let (_, depth) = blocktree::find_mut(&mut self.tree, block_hash)
+            .ok_or_else(|| BlockDoesNotExtendTree(block_hash.clone()))?;
+        Ok(depth)
+    }
+
+    // Inserts the block hash of the block that should be received.
+    pub fn insert_next_block_hash(
+        &mut self,
+        prev_block_hash: &BlockHash,
+        block_hash: &BlockHash,
+        stable_height: Height,
+    ) {
+        let height = match self.next_block_hashes.get_height(prev_block_hash) {
+            Some(prev_height) => *prev_height,
+            None => {
+                if let Ok(depth) = self.block_depth(prev_block_hash) {
+                    stable_height + depth
+                } else {
+                    return;
```

Comment:
It looks like we're silently ignoring the case where we are inserting a block hash that doesn't extend the tree. I think it's more clear for this API to return a `BlockDoesNotExtendTree` error and let the caller decide what to do.